x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/cache/CacheService.java

```diff
@@ -338,82 +351,130 @@ public void removeFromCache(final CacheKey cacheKey) {
+            if (allowShardsEvictions) {
+                final ShardEviction shardEviction = new ShardEviction(snapshotUUID, snapshotIndexName, shardId);
+                pendingShardsEvictions.computeIfAbsent(shardEviction, shard -> threadPool.generic().submit(new AbstractRunnable() {
+                    @Override
+                    protected void doRun() {
+                        processShardEviction(shardEviction);
+                    }
 
-                @Override
-                public void onFailure(Exception e) {
-                    assert false : e;
-                    logger.warn(
-                        () -> new ParameterizedMessage("failed to evict cache files associated with evicted shard {}", shardEviction),
-                        e
-                    );
-                }
-            });
+                    @Override
+                    public void onFailure(Exception e) {
+                        logger.warn(
+                            () -> new ParameterizedMessage("failed to evict cache files associated with shard {}", shardEviction),
+                            e
+                        );
+                        assert false : e;
+                    }
+                }));
+            }
         }
     }
 
     /**
-     * Allows to run the specified {@link Runnable} if the shard represented by the triplet ({@link SnapshotId}, {@link IndexId},
-     * {@link SnapshotId}) is still marked as evicted at the time this method is executed. The @link Runnable} will be executed
-     * while the current thread is holding the lock associated to the shard.
+     * Waits for the cache files associated with the shard represented by ({@link SnapshotId}, {@link IndexId}, {@link SnapshotId}) to be
+     * evicted if the shard is marked as evicted in cache at the time this method is executed.
      *
-     * @param snapshotId the snapshot the evicted searchable snapshots shard belongs to
-     * @param indexId    the index in the snapshot the evicted searchable snapshots shard belongs to
-     * @param shardId    the searchable snapshots shard id
-     * @param runnable   a runnable to execute
+     * @param snapshotUUID      the snapshot's unique identifier
+     * @param snapshotIndexName the name of the index in the snapshot
+     * @param shardId           the {@link ShardId}
      */
-    public void runIfShardMarkedAsEvictedInCache(SnapshotId snapshotId, IndexId indexId, ShardId shardId, Runnable runnable) {
-        runIfShardMarkedAsEvictedInCache(new ShardEviction(snapshotId.getUUID(), indexId.getName(), shardId), runnable);
+    public void waitForCacheFilesEvictionIfNeeded(String snapshotUUID, String snapshotIndexName, ShardId shardId) {
+        final Future<?> future;
+        synchronized (shardsEvictionsMutex) {
+            future = pendingShardsEvictions.get(new ShardEviction(snapshotUUID, snapshotIndexName, shardId));
+            if (future == null) {
+                return;
```

Comment:
I think we need to check if stopped here (or if we allow shard evictions) and if not throw an exception (`AlreadyClosedException`?) to terminate any concurrent directories doing `loadSnapshot`.

Otherwise, [skipping shard evictions](https://github.com/elastic/elasticsearch/pull/67160/files#diff-6dbbb92e239f75e947fb77603974dc99d62f2e54af9b4bbb376c42142656c697R360) seems unsafe, if a directory is about to `loadSnapshot` concurrently?

Reply:
:open_mouth: sorry, I thought I've done this but apparently not. I agree we need to throw an exception here.

Reply:
I think we need the exception regardless of whether it is in `pendingShardsEvictions` or not, since we may have neglected to add an eviction in `markShardAsEvictedInCache` when `allowShardsEvctions == false`.

Reply:
My take on this was to let the directory fails itself as opening or listing files should fail if the directory is closed. 

But since shards are closed after lifecycle components are stopped I think you are right and we should always throw here, otherwise there is a chance that an empty cache file got used and throws an `EOFException`.

I pushed 0a0b260571a19cf4774ac0ffdb35b57870c7a000